### PR TITLE
unix, windows: add uv_fileno

### PR DESCRIPTION
**Pull request by saghul**
Returns the platform specific file descriptor for handles that are
backed by one. The datatype is abstracted as uv_os_fd_t, which maps to
int on Unices and HANDLE on Windows.

Users can use this function to set specific socket options, for example,
in a non portable way.

This function is essentially a shotgun, you better be careful with
whatever you do with it, don't blame me if you used it to get the fd of
a stream, close it yourself and expect things to Just Work.

### CODE REVIEW

**Comment by saghul** [comment]
This patch is unifinished (Windows part and tests are missing), but I wanted to start a discussion about it. I can't think of a way which would please everyone while keeping our fds hidden, so I think it might not be that bad to expose them for those with legit use cases such as setting some socket options here and there, we can't possibly map all of them.

About the Windows side, HANDLE is just a void*, which means we wouldn't be able to cast it to int because of 64bit systems. However, reading around seems like only the lower 32 bits are used because a process can only have 2^24 handles in that handle table. Anyway, I went for the typedef, since this is a platform dependent thing and feels more future-proof.

Thoughts?

Calling the cavalry! @indutny, @bnoordhuis, @piscisaureus

**Comment by txdv** [comment]
IMO this is a good idea.

Let the user do whatever he wants.

**Comment by txdv** [comment]
This is such an easy fix, why didn't we have it earlier?

**Comment by saghul** [comment]
Just like with scotch, some things take time to become better, I guess :-P

File: include/uv.h [hunk 1]
```diff
@@ -615,6 +615,9 @@ UV_EXTERN void uv_close(uv_handle_t* handle, uv_close_cb close_cb);
  */
 UV_EXTERN int uv_recv_buffer_size(uv_handle_t* handle, int* value);
 
+/* TODO */
+UV_EXTERN uv_os_fd_t uv_fileno(uv_handle_t* handle);
```
Comment (bnoordhuis) [inline include/uv.h]: Maybe `UV_EXTERN int uv_file(const uv_handle_t* handle, uv_os_fd_t* fd);`?  Libuv can't really return a meaningful error now.

File: src/unix/core.c [hunk 2]
```diff
@@ -662,6 +662,25 @@ void uv_disable_stdio_inheritance(void) {
 }
 
 
+uv_os_fd_t uv_fileno(uv_handle_t *handle) {
+  switch (handle->type) {
+  case UV_TCP:
+  case UV_NAMED_PIPE:
+  case UV_TTY:
+    return ((uv_stream_t *) handle)->io_watcher.fd;
```
Comment (bnoordhuis) [inline src/unix/core.c]: I believe that on OS X, io_watcher.fd is not always the actual file descriptor.  Also, you have a style error a few lines up.  :-)

File: include/uv.h (hunk 1 above)
Comment (saghul) [inline include/uv.h]: That crossed my mind, but in the end I thought that I'd just return -1 when there is no fd. We could give EINVAL if you pass an unsupported handle type, I guess.

File: src/unix/core.c (hunk 2 above)
Comment (saghul) [inline src/unix/core.c]: I vaguely remember that, I'll check it out, thanks!. Based on your comments, I'll assume you are on board with this then :-)

File: include/uv.h (hunk 1 above)
Comment (bnoordhuis) [inline include/uv.h]: But how are you going to do that on Windows?  HANDLE is a typedef for `void*`.

Comment (saghul) [inline include/uv.h]: I guess the same way Windows does for INVALID_SOCKET :-P On a more serious note, I do like your proposal, I'll change this.

Comment (saghul) [inline include/uv.h]: FWIW, Windows defines INVALID_SOCKET as SOCKET(~0).

**Comment by saghul** [comment]
Updated the API and Unix implementation as per Ben's suggestions. It returns EINVAL if the handle is not a stream, poll or udp and EBADF if the fd is -1.

File: src/unix/core.c [hunk 3]
```diff
@@ -662,6 +662,32 @@ void uv_disable_stdio_inheritance(void) {
...
+  case UV_POLL:
+    *fd = ((uv_poll_t *) handle)->io_watcher.fd;
+    break;
+
+  default:
+    return -EINVAL;
+  }
+
+  if (*fd == -1)
+    return -EBADF;
```
Comment (bnoordhuis) [inline src/unix/core.c:688]: It's kind of bad style to return an error code but still mutate the out parameter, IMO.

Comment (saghul) [inline src/unix/core.c:688]: True that, will fix.

**Comment by saghul** [comment]
Added tests and Windows implementation. The TTY test fails, I need to open $conin it seems, like the TTY test does, will do.

File: test/test-handle-fileno.c [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+  ASSERT(r == 0);
+  r = uv_fileno((uv_handle_t*) &pipe, &fd);
+  ASSERT(r == UV_EBADF);
+  r = uv_pipe_bind(&pipe, TEST_PIPENAME);
+  ASSERT(r == 0);
+  r = uv_fileno((uv_handle_t*) &pipe, &fd);
+  ASSERT(r == 0);
+  uv_close((uv_handle_t*) &pipe, NULL);
+#ifndef _WIN32
+  /* close is not immediate on Windows */
```
Comment (bnoordhuis) [inline test/test-handle-fileno.c]: Can't you filter on UV__HANDLE_CLOSING in uv_fileno()?

File: include/uv.h [hunk 5]
```diff
@@ -615,6 +615,8 @@ UV_EXTERN void uv_close(uv_handle_t* handle, uv_close_cb close_cb);
  */
 UV_EXTERN int uv_recv_buffer_size(uv_handle_t* handle, int* value);
 
+UV_EXTERN int uv_fileno(uv_handle_t* handle, uv_os_fd_t* fd);
```
Comment (bnoordhuis) [inline include/uv.h]: I think handle can be const?

File: test/test-handle-fileno.c (hunk 4 above)
Comment (saghul) [inline test/test-handle-fileno.c]: I'm not sure what guarantees we make about it. The handle (libuv wise) is closing, but the actual Windows HANDLE might need to stick around for a bit more.

File: include/uv.h (hunk 5 above)
Comment (saghul) [inline include/uv.h]: I think so, yes. Good call!

**Comment by saghul** [comment]
Tests pass on Windows now.

**Comment by saghul** [comment]
@bnoordhuis thanks, const-ified the handle now.

**Comment by saghul** [comment]
Any further comments on this?

**Comment by txdv** [comment]
Why do we set it to invalid?

**Comment by saghul** [comment]
> Why do we set it to invalid?

Because otherwise you get it with uv_fileno after you have closed it so it might belong to another handle that reused it. This wasn't a problem before, because it was not exposed, but now that people get access to file descriptors, we cannot give them stalled ones.

Another possibility might be to return EBADF if uv_is_closing(handle). That might actually simplify things a bit...

**Comment by saghul** [comment]
I'll give that a shot (I think that's what @bnoordhuis said before, but I didn't quite get it until now). I'd still leave the commit that sets handles to invalid, as a defense mechanism.

**Comment by saghul** [comment]
Updated with uv_is_closing check, now we no longer need any ifdef-ery on the tests \o/.

**Comment by bnoordhuis** [comment]
LGTM.  And yes, that was what I meant. :-)

**Comment by saghul** [comment]
Thanks folks, landed! Load your shotguns.

**Pull request merged**